### PR TITLE
sink(ticdc): set max-message-bytes default to 10m

### LINKED ISSUE #4041: TiCDC sink initialization should check topic or broker's max message bytes configuration for all kinds of protocols.
### What did you do?

In v4.0.16, if use non `open-protocol`, such as `canal-json`, `canal`, `avro` to initialize the Kafka sink, changefeed may meet `message too large` error from the Sarama producer.

Furthermore, the TiCDC cluster was upgraded from v4.0.14 to v4.0.16, we should also consider upgrade compatibility, which means old existing changefeeds should boost and run normal.

### What did you expect to see?

changefeed works normally, no matter which kind of protocol.

### What did you see instead?

```
[CDC:ErrKafkaAsyncSendMessage]kafka: Failed to produce message to topic cdc-eps_rt_ods-ods_market_flw_lx_app_log_prod: kafka server: Message was too large, server rejected it to avoid allocation error."]
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

TiCDC version (execute `cdc version`):

```console
v4.0.16
```

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve? 
close #4041
set Kafka producer's default `max-message-bytes` to 10M, and use the minimum value among sink-uri's `max-message-bytes`, broker's `message.max.bytes`, topic's `max.message.bytes` to initialize the producer.

### What is changed and how it works?
* set the default value of `max-message-bytes` to 10M
* use the min value of the 3 mentioned above to initialize the producer.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### Release note 

```release-note
Set `max-message-bytes` default to 10M, and use the min value with topic and broker to initialize the producer.
```

### CODE REVIEW

File: cdc/sink/codec/json.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ const (
 	// BatchVersion1 represents the version of batch format
 	BatchVersion1 uint64 = 1
 	// DefaultMaxMessageBytes sets the default value for max-message-bytes
-	DefaultMaxMessageBytes int = 1 * 1024 * 1024 // 1M
+	DefaultMaxMessageBytes int = 10 * 1024 * 1024 // 1M
```
Comment (amyangfei) [inline cdc/sink/codec/json.go:41]:
```suggestion
	DefaultMaxMessageBytes int = 10 * 1024 * 1024 // 10M
```

File: cdc/sink/producer/kafka/kafka.go [hunk 2]
```diff
@@ -354,17 +354,18 @@ func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic
...
 		}
+
 		if topicMaxMessageBytes < config.MaxMessageBytes {
-			return cerror.ErrKafkaInvalidConfig.GenWithStack(
-				"topic already exist, and topic's max.message.bytes(%d) less than max-message-bytes(%d). "+
-					"Please make sure `max-message-bytes` not greater than topic's `max.message.bytes`",
-				topicMaxMessageBytes, config.MaxMessageBytes)
+			log.Warn("topic's `max.message.bytes` less than user set `max-message-bytes`,"+
+				"use topic's `max.message.bytes` to initialize the Kafka producer",
+				zap.Int("max.message.bytes", topicMaxMessageBytes),
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka.go:366]:
```suggestion
				zap.Int("message.max.bytes", topicMaxMessageBytes),
```

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:366]: for the topic, it's `max.message.bytes`, for the broker, it's `message.max.bytes`

File: cdc/sink/codec/json.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ const (
 	// BatchVersion1 represents the version of batch format
 	BatchVersion1 uint64 = 1
 	// DefaultMaxMessageBytes sets the default value for max-message-bytes
-	DefaultMaxMessageBytes int = 1 * 1024 * 1024 // 1M
+	DefaultMaxMessageBytes int = 10 * 1024 * 1024 // 10M
```
Comment (amyangfei) [inline cdc/sink/codec/json.go:41]: Can we move this constant value to pkg/config and use one const only.

File: cdc/sink/producer/kafka/kafka.go (hunk 2 above)
Comment (amyangfei) [reply cdc/sink/producer/kafka/kafka.go:366]: Got it, I found it in Kafka document, it's strange.

File: cdc/sink/producer/kafka/kafka_test.go [hunk 4]
```diff
@@ -222,46 +231,49 @@ func (s *kafkaSuite) TestValidateMaxMessageBytesAndCreateTopic(c *check.C) {
...
 	detail := &sarama.TopicDetail{
 		NumPartitions: 3,
 		// Does not contain max message bytes information.
 		ConfigEntries: make(map[string]*string),
 	}
 	err = adminClient.CreateTopic("test-topic", detail, false)
 	c.Assert(err, check.IsNil)
-	err = validateMaxMessageBytesAndCreateTopic(adminClient, "test-topic", config)
+	err = validateMaxMessageBytesAndCreateTopic(adminClient, "test-topic", config, cfg)
 	c.Assert(err, check.IsNil)
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka_test.go:266]: We can check the value of `cfg.Producer.MaxMessageBytes` here.

**Review by amyangfei — COMMENTED** [review]
rest lgtm

**Review by amyangfei — APPROVED** [review]

File: pkg/config/sink.go [hunk 5]
```diff
@@ -21,6 +21,9 @@ import (
 	cerror "github.com/pingcap/tiflow/pkg/errors"
 )
 
+// DefaultMaxMessageBytes sets the default value for max-message-bytes
+var DefaultMaxMessageBytes int = 10 * 1024 * 1024 // 10M
```
Comment (hi-rustin) [inline pkg/config/sink.go:25]:
```suggestion
const DefaultMaxMessageBytes  = 10 * 1024 * 1024 // 10M
```

File: cdc/sink/producer/kafka/kafka.go [hunk 6]
```diff
@@ -345,7 +345,7 @@ func kafkaClientID(role, captureAddr, changefeedID, configuredClientID string) (
 	return
 }
 
-func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic string, config *Config) error {
+func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic string, config *Config, saramaConfig *sarama.Config) error {
```
Comment (hi-rustin) [inline cdc/sink/producer/kafka/kafka.go:348]: This func seems strange at the moment. Can we split it up?
1. getMaxMessageBytes is responsible for getting the right value
2. createTopic is used to create the topic

This way we won't have to pass two configurations and modify them at the same time. Now this function has become very complicated. Originally we didn't modify Sarama's configuration in this method. Now it is not only responsible for creating the topic, but also for setting the sarma configuration correctly, but this configuration is not directly related to creating the topic. It does too many things. I prefer to keep them separate.

Comment (hi-rustin) [reply cdc/sink/producer/kafka/kafka.go:348]: If there is a better way to get it not to set it up at the same time that would work too. I can only think of separating it at the moment.
The sarma configuration has been patched once above via kafka's configuration, but we're modifying it at the same time in this function.

It is currently confusing from either the caller's or the test code's point of view.

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:348]: This worthy another PR to do it.

Comment (hi-rustin) [reply cdc/sink/producer/kafka/kafka.go:348]: I think it might be better to do it in this time. It should just be a simple split method will work.

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:348]: I would leave that to another PR, as a future refinement.

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:348]: I still think it would be better to do this after the release, to prevent some potential new problems.

**Review by overvenus — APPROVED** [review]

**Pull request merged**